File: backend/services/stigmer-server/pkg/domain/mcpserver/controller/disconnect_oauth.go
package mcpserver

import (
	"context"

	"github.com/rs/zerolog/log"
	mcpserverv1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/agentic/mcpserver/v1"
	grpclib "github.com/stigmer/stigmer/backend/libs/go/grpc"
)

// DisconnectOAuth tears down a user's OAuth connection for an MCP server.
//
// Deletes the OAuthGrant record and its associated managed environment
// (which holds the access and refresh tokens). The MCP server definition
// is unchanged — only the caller's personal OAuth connection is removed.
//
// Idempotent: if no grant exists for the (caller, resource_id, org) tuple,
// returns disconnected=false without error. This supports race conditions,
// retries after partial failures, and desired-state semantics.
//
// Delete order: managed environment first (eliminates secrets), then grant
// record (metadata only). If grant deletion fails after environment deletion,
// the orphaned grant is harmless metadata pointing to a deleted environment.
func (c *McpServerController) DisconnectOAuth(
	ctx context.Context,
	input *mcpserverv1.DisconnectOAuthInput,
) (*mcpserverv1.DisconnectOAuthOutput, error) {
	if c.oauthGrantStore == nil {
		return &mcpserverv1.DisconnectOAuthOutput{Disconnected: false}, nil
	}

	resourceID := input.GetResourceId()
	if resourceID == "" {
		return nil, grpclib.InvalidArgumentError("resource_id is required")
	}
	org := input.GetOrg()
	if org == "" {
		return nil, grpclib.InvalidArgumentError("org is required")
	}

	// OSS mode: single user, empty identity_account_id.
	grant, err := c.oauthGrantStore.Find(ctx, "", resourceID, org)
	if err != nil {
		return nil, grpclib.InternalError(err, "failed to look up OAuth grant")
	}

	if grant == nil {
		log.Debug().
			Str("resource_id", resourceID).
			Str("org", org).
			Msg("No OAuth grant to disconnect")
		return &mcpserverv1.DisconnectOAuthOutput{Disconnected: false}, nil
	}

	envID := grant.EnvironmentID
	if envID != "" && c.managedEnvService != nil {
		if err := c.managedEnvService.DeleteManagedEnvironment(ctx, envID); err != nil {
			log.Warn().Err(err).
				Str("resource_id", resourceID).
				Str("environment_id", envID).
				Msg("Failed to delete managed environment — may already be deleted")
		}
	}

	if err := c.oauthGrantStore.Delete(ctx, "", resourceID, org); err != nil {
		return nil, grpclib.InternalError(err, "failed to delete OAuth grant")
	}

	log.Info().
		Str("resource_id", resourceID).
		Str("org", org).
		Bool("env_deleted", envID != "").
		Msg("OAuth connection disconnected")

	return &mcpserverv1.DisconnectOAuthOutput{Disconnected: true}, nil
}

File: backend/services/stigmer-server/src/domain/mcpserver/disconnect-oauth.ts
/**
 * disconnectOAuth — ports
 * pkg/domain/mcpserver/controller/disconnect_oauth.go: tear down a user's
 * OAuth connection for an MCP server. Deletes the OAuthGrant record and
 * its associated managed environment (which holds the access and refresh
 * tokens); the MCP server definition is unchanged.
 *
 * Idempotent: no grant for the (caller, resource_id, org) tuple returns
 * disconnected=false without error — race conditions, retries after
 * partial failures, and desired-state semantics all rely on it.
 *
 * Delete order: managed environment first (eliminates secrets), then
 * grant record (metadata only). If grant deletion fails after environment
 * deletion, the orphaned grant is harmless metadata pointing to a deleted
 * environment.
 *
 * Proven by mcpserver-oauth.conformance.test.ts (guards + no-grant
 * idempotence, CONFORMANCE_TARGET=local) and
 * mcpserver-connect.conformance.test.ts (teardown,
 * CONFORMANCE_TARGET=local-execution).
 */
import { create } from "@bufbuild/protobuf";

import type {
  DisconnectOAuthInput,
  DisconnectOAuthOutput,
} from "@stigmer/protos/ai/stigmer/agentic/mcpserver/v1/io_pb";
import { DisconnectOAuthOutputSchema } from "@stigmer/protos/ai/stigmer/agentic/mcpserver/v1/io_pb";

import { McpServerCommandController } from "@stigmer/protos/ai/stigmer/agentic/mcpserver/v1/command_pb";

import type { CallerIdentity } from "../../extensions/identity.js";
import { internalError, invalidArgumentError } from "../../pipeline/errors.js";
import { authorizeDirect } from "../../pipeline/steps/authorize.js";
import type { McpServerConnectDeps } from "./connect.js";

export async function disconnectOAuth(
  deps: McpServerConnectDeps,
  input: DisconnectOAuthInput,
  identity: CallerIdentity,
): Promise<DisconnectOAuthOutput> {
  const resourceId = input.resourceId;
  if (resourceId === "") {
    throw invalidArgumentError("resource_id is required");
  }
  const org = input.org;
  if (org === "") {
    throw invalidArgumentError("org is required");
  }
  // The annotation's can_connect check (validate → authorize, the Java
  // McpServerDisconnectOAuthHandler order — no load step; on the
  // multi-tenant edition an unresolvable id answers through the
  // authorizer's ruled uniform posture). C2 Stage 4.
  await authorizeDirect(
    McpServerCommandController.method.disconnectOAuth,
    deps.authorizer,
    identity,
    input,
  );

  // OSS mode: single user, empty identity_account_id.
  let grant;
  try {
    grant = await deps.oauthGrants.find("", resourceId, org);
  } catch (error) {
    throw internalError(error, "failed to look up OAuth grant");
  }

  if (grant === undefined) {
    deps.logger.debug("No OAuth grant to disconnect", {
      resource_id: resourceId,
      org,
    });
    return create(DisconnectOAuthOutputSchema, { disconnected: false });
  }

  const envId = grant.environmentId;
  if (envId !== "") {
    try {
      await deps.managedEnv.deleteManagedEnvironment(envId);
    } catch (error) {
      deps.logger.warn(
        "Failed to delete managed environment — may already be deleted",
        {
          resource_id: resourceId,
          environment_id: envId,
          error: error instanceof Error ? error.message : String(error),
        },
      );
    }
  }

  try {
    await deps.oauthGrants.delete("", resourceId, org);
  } catch (error) {
    throw internalError(error, "failed to delete OAuth grant");
  }

  deps.logger.info("OAuth connection disconnected", {
    resource_id: resourceId,
    org,
    env_deleted: envId !== "",
  });

  return create(DisconnectOAuthOutputSchema, { disconnected: true });
}
